Scrape a Mesos Master process

wndhydrnt opened the pull request:
This PR adds support for scraping metrics exposed by a Mesos Master process.

It also introduces a change to the CLI flags by specifying `exporter.scrape-mode` to distinguish between the different modes that mesos_exporter supports. Also the different flags that included the term `url` have been unified into the flag `exporter.url`.

For now, I have added only those metrics also exposed by wndhydrnt/mesos-task-exporter (which I'd like to deprecate in favour of this repo) as these are the metrics I use every day.

Together with prometheus/prometheus#1027, these changes make it possible to monitor a Mesos cluster by running an instance of mesos_exporter next to a Mesos process.

wndhydrnt commented:
PTAL @antonlindstrom @juliusv @tommyulfsparre

brian-brazil commented on README.md:
```diff
@@ -13,23 +13,53 @@ Name                           | Description
+
+##### Example
+
+    ./mesos_exporter -exporter.interval=15s -exporter.scrape-mode=master -exporter.url=http://127.0.0.1:5050
```
You should only scrape the master when prometheus scrapes you, so there shouldn't be an interval configuration option

wndhydrnt commented:
I forgot one feature in the initial push. The name of the task and the name of the framework that spawned the task are now added as labels to each metric generated from the `/monitor/statistics.json` endpoint.
The names are the same that are visible in the Mesos UI.
This simplifies writing expressions that target several tasks of the same app as the names usually do not change while IDs do.

StephanErb commented:
What's the status of this pull request? Would love to get to use those new metrics.

juliusv commented:
Good question. /cc @wndhydrnt @brian-brazil

juliusv commented on main.go:
```diff
@@ -20,16 +20,16 @@ const concurrentFetch = 100
-	metricsPath    = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics")
-	scrapeInterval = flag.Duration("exporter.interval", (60 * time.Second), "Scrape interval duration")
+	addr                 = flag.String("web.listen-address", ":9105", "Address to listen on for web interface and telemetry")
+	autoDiscoverInterval = flag.Duration("exporter.discover-interval", (10 * time.Minute), "Interval at which to update available slaves from a Mesos Master. Only used if exporter.scrape-mode=discover.")
```
No parens around `(10 * time.Minute)` needed here.

juliusv commented on main.go:
```diff
@@ -20,16 +20,16 @@ const concurrentFetch = 100
+	autoDiscoverInterval = flag.Duration("exporter.discover-interval", (10 * time.Minute), "Interval at which to update available slaves from a Mesos Master. Only used if exporter.scrape-mode=discover.")
+	queryURL             = flag.String("exporter.url", "", "The URL of a Mesos Slave, if mode=slave. The URL of a Mesos Master, if mode=discover or mode=master.")
+	metricsPath          = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics")
+	scrapeInterval       = flag.Duration("exporter.interval", (60 * time.Second), "Scrape interval duration")
```
Same.

juliusv commented on main.go:
```diff
@@ -56,6 +56,135 @@ var (
+
+	frameworkLabels = []string{"id", "name"}
+
+	frameworkResourcesUsedCpus = prometheus.NewDesc(
```
`...Cpus` -> `CPUs` (https://github.com/golang/go/wiki/CodeReviewComments#initialisms)

juliusv commented on main.go:
```diff
@@ -56,6 +56,135 @@ var (
+
+	frameworkResourcesUsedCpus = prometheus.NewDesc(
+		"mesos_framework_resources_used_cpus",
+		"Fractional CPUs used by a framework",
```
For consistency with all other help strings, end them with a period.

juliusv commented on main.go:
```diff
@@ -56,6 +56,135 @@ var (
+	)
+
+	frameworkResourcesUsedDisk = prometheus.NewDesc(
+		"mesos_framework_resources_used_disk",
```
This needs a unit suffix (`_bytes`).

juliusv commented on main.go:
```diff
@@ -56,6 +56,135 @@ var (
+	)
+
+	frameworkResourcesUsedMemory = prometheus.NewDesc(
+		"mesos_framework_resources_used_memory",
```
This also needs a unit suffix (`_bytes`).

juliusv commented on main.go:
```diff
@@ -56,6 +56,135 @@ var (
+	slaveMetrics = []snapshotMetric{
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_slave_cpus_total",
```
`_total` is used for counter metrics, so let's just call this `mesos_slave_cpus`.

juliusv commented on main.go:
```diff
@@ -56,6 +56,135 @@ var (
+		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_slave_disk_total",
```
Same here, and also add a unit suffix: `mesos_slave_disk_bytes` or `mesos_slave_total_disk_bytes` (if that's clearer, as long as there's no `_total` as a suffix).

juliusv commented on main.go:
```diff
@@ -56,6 +56,135 @@ var (
+		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_slave_memory_total",
```
Same here: `mesos_slave_memory_bytes`

juliusv commented on main.go:
```diff
@@ -63,24 +192,80 @@ var httpClient = http.Client{
 }
 
+type resources struct {
+	Cpus float64
```
`Cpus` -> `CPUs`, same for `Pid` -> `PID`, `Id` -> `ID`, `FrameworkId` -> `FrameworkID`.

juliusv replied:
(it shouldn't affect the JSON unmarshaling)

juliusv commented on main.go:
```diff
@@ -93,33 +278,44 @@ func newMesosExporter(opts *exporterOpts) *periodicExporter {
+		go runEvery(e.scrapeSlaves, e.opts.interval)
+	case "master":
+		log.Info("starting mesos_exporter in scrape mode 'master'")
+		e.queryURL = parseMasterUrl(opts.queryURL)
```
`parseMasterUrl` -> `parseMasterURL`

juliusv commented on main.go:
```diff
@@ -153,52 +356,91 @@ func (e *periodicExporter) fetch(urlChan <-chan string, metricsChan chan<- prome
+		var state slaveState
+		stateURL := fmt.Sprintf("%s/state.json", u)
+
+		err = getJson(&state, stateURL)
```
`getJson` -> `getJSON`

juliusv commented on main.go:
```diff
@@ -153,52 +356,91 @@ func (e *periodicExporter) fetch(urlChan <-chan string, metricsChan chan<- prome
 
-		monitorURL := fmt.Sprintf("%s/monitor/statistics.json", u)
-		resp, err := httpClient.Get(monitorURL)
+		taskInfo := make(map[string]exporterTaskInfo)
```
Avoid `make` unless it's necessary: `taskInfo := map[string]exporterTaskInfo{}`

juliusv commented on main.go:
```diff
@@ -221,6 +463,68 @@ func (e *periodicExporter) setMetrics(ch chan prometheus.Metric) {
+		return
+	}
+
+	metrics := make([]prometheus.Metric, 0)
```
Same here, just: `metrics := []prometheus.Metric{}` unless you're going to preallocate a slice with a non-zero length (which you could do here theoretically, `len(state.Frameworks) * 3`, but maybe not worth it).

wndhydrnt commented:
@juliusv thanks for the review. I've updated the code according to your suggestions.

juliusv commented:
@wndhydrnt Awesome, thanks! Taking another look!

juliusv commented on main.go:
```diff
@@ -56,6 +56,135 @@ var (
+	)
+
+	frameworkResourcesUsedDisk = prometheus.NewDesc(
+		"mesos_framework_resources_used_disk_megabytes",
```
Let's keep this in base units (`_bytes`). Easier to correlate with disk size metrics from other exporters which are all in bytes, plus otherwise it's not 100% clear whether it's MiB or MB otherwise. Probably MiB here, so multiply by 1024^2?

Same for the one below.

juliusv commented:
Looks :+1: to me otherwise.

wndhydrnt commented:
@juliusv addressed your comment regarding `_bytes`. As the `/metrics/snapshot` endpoint also exposes megabytes instead of bytes, I've added the functionality to convert the value of a metric by passing a function to the `snapshotMetric` struct.

Let me know if this is fine with you and I'll squash the commits.

juliusv commented:
@wndhydrnt Thanks!

Hmm, I'm wondering if adding two extra constructor functions isn't a bit of over-engineering for that. How about still keeping all of these metric definitions as struct literals instead of function calls and simply have an optional `convertFn` field that can be `nil` (when omitted), and it's only applied if it's not `nil`?

Also, the `convertMegabytesToBytes` function is so small that you could consider inlining it, though that's up to taste:

``` go
  // ...
  convertFn: func(v float64) { return v * 1024 * 1024 },
  // ...
```

If you keep it, I'd just name it `megabytesToBytes`. Seems to be more common to omit prefixes like `get`, `compute`, `convert`, etc. from method names if they are already clear without it.

wndhydrnt commented:
@juliusv addressed your last comment

juliusv commented:
:+1: Awesome, please squash and I'll merge!

wndhydrnt commented:
Squashed

juliusv commented:
Thanks!
